gnd/src/main/java/com/google/android/gnd/repository/FeatureRepository.java

```diff
@@ -0,0 +1,126 @@
+import io.reactivex.Completable;
+import io.reactivex.Flowable;
+import io.reactivex.Maybe;
+import java8.util.Optional;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Coordinates persistence and retrieval of {@link Feature} instances from remote, local, and in
+ * memory data stores. For more details on this pattern and overall architecture, see
+ * https://developer.android.com/jetpack/docs/guide.
+ */
+@Singleton
+public class FeatureRepository {
+  private final LocalDataStore localDataStore;
+  private final ProjectRepository projectRepository;
+  private final DataSyncWorkManager dataSyncWorkManager;
+
+  @Inject
+  public FeatureRepository(
+      LocalDataStore localDataStore,
+      RemoteDataStore remoteDataStore,
+      ProjectRepository projectRepository,
+      DataSyncWorkManager dataSyncWorkManager) {
+    this.localDataStore = localDataStore;
+    this.projectRepository = projectRepository;
+    this.dataSyncWorkManager = dataSyncWorkManager;
+
+    streamFeaturesToLocalDb(remoteDataStore);
+  }
+
+  /**
+   * Mirrors features in the current project from the remote db into the local db when the network
+   * is available. When invoked, will first attempt to resync all features from the remote db,
+   * subsequently syncing only remote changes.
+   */
+  private void streamFeaturesToLocalDb(RemoteDataStore remoteDataStore) {
+    // TODO: Move to Application or background service.
+    // TODO: Is this even working? If the returned Disposable is garbage collected this will be
+    // interrupted.
+    projectRepository
+        .getActiveProjectOnceAndStream()
+        .compose(Loadable::values)
+        .switchMap(remoteDataStore::loadFeaturesOnceAndStreamChanges)
+        .switchMap(event -> updateLocalFeature(event).toFlowable())
+        .subscribe();
+  }
+
+  private Completable updateLocalFeature(RemoteDataEvent<Feature> event) {
+    switch (event.getEventType()) {
+      case ENTITY_LOADED:
+      case ENTITY_MODIFIED:
+        return event.value().map(localDataStore::mergeFeature).orElse(Completable.complete());
+      case ENTITY_REMOVED:
+        // TODO: Delete features:
+        // localDataStore.removeFeature(event.getEntityId());
+        return Completable.complete();
+      case ERROR:
+        return Completable.error(event.error().get());
+      default:
+        return Completable.error(
+            new UnsupportedOperationException("Event type: " + event.getEventType()));
+    }
+  }
+
+  // TODO: Only return feature fields needed to render features on map.
+  // TODO(#127): Decouple from Project and accept id instead.
+  public Flowable<ImmutableSet<Feature>> getFeaturesOnceAndStream(Project project) {
+    return localDataStore.getFeaturesOnceAndStream(project);
+  }
+
+  // TODO(#127): Decouple Project from Feature and remove projectId.
+  // TODO: Replace with Single and treat missing feature as error.
+  public Maybe<Feature> getFeature(String projectId, String featureId) {
```

Comment:
If my reading is accurate, I think the impl is slightly misleading here. The signature would suggest that if I know project P exists and feature F exists, getFeature(P, F) would return F, however, this method actually has the additional requirement that P === ActiveProject P* if that doesn't hold, even if P and F exist (P,F) returns nothing.

Reply:
Yep, this is a known limitation. In practice features will only ever be loaded for the active project so this isn't an issue. In either case fixing it here is out of the scope of this PR. Added a TODO to fix it going forward.